src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -232,8 +246,79 @@ private struct MethodTable
             // Ideally this would just be a single dereference:
             // mov tmp, qword ptr [rax] ; rax = obj ref, tmp = MethodTable* pointer
         }
+    }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern object GetUninitializedObjectInternal(Type type);
+    // Helper class to assist with unsafe pinning of arbitrary objects.
+    // It's used by VM code.
+    internal class RawData
+    {
+        public byte Data;
+    }
+
+    internal class RawArrayData
+    {
+        public uint Length; // Array._numComponents padded to IntPtr
+#if BIT64
+        public uint Padding;
+#endif
+        public byte Data;
+    }
+
+    // Subset of src\vm\methodtable.h
+    [StructLayout(LayoutKind.Explicit)]
+    internal unsafe struct MethodTable
+    {
+        [FieldOffset(0)]
+        public ushort ComponentSize;
+        [FieldOffset(0)]
+        private uint Flags;
+        [FieldOffset(4)]
+        public uint BaseSize;
+
+        // WFLAGS_HIGH_ENUM
+        private const uint enum_flag_ContainsPointers = 0x01000000;
+        // private const uint enum_flag_HasComponentSize = 0x80000000;
+
+        public bool HasComponentSize
+        {
+            get
+            {
+                // return (Flags & enum_flag_HasComponentSize) != 0;
+
+                // The MethodTable::Flags field will have its high bit set if the
+                // type has component size. See member
+                // MethodTable:IsStringOrArray in src\vm\methodtable.h for full details.
+                return (int)Flags < 0;
+            }
+        }
+
+        public bool ContainsGCPointers
+        {
+            get
+            {
+                return (Flags & enum_flag_ContainsPointers) != 0;
+            }
+        }
+
+        public bool IsMultiDimensionalArray
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                Debug.Assert(HasComponentSize);
+                return BaseSize > (uint)(3 * sizeof(IntPtr));
+            }
+        }
+
+        // Returns rank of multi-dimensional array rank, 0 for sz arrays
+        public int MultiDimensionalArrayRank
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                Debug.Assert(HasComponentSize);
+                return (int)((BaseSize - (uint)(3 * sizeof(IntPtr))) / (uint)(2 * sizeof(int)));
```

Comment:
I don't know how to comment this logic, but I think it should be here. The 3 pointer sizes are for the actual raw object layout right?